domain/consensus/processes/consensusstatemanager/resolve.go

```diff
@@ -49,55 +44,136 @@ func (csm *consensusStateManager) ResolveVirtual(maxBlocksToResolve uint64) (*ex
-			log.Debugf("Has more than %d blocks to resolve. Changing the resolve tip to %s", maxBlocksToResolve, resolveTip)
+		if status == externalapi.StatusUTXOValid || status == externalapi.StatusUTXOPendingVerification {
+			return tip, status, nil
 		}
+	}
 
-		blockStatus, reversalData, err := csm.resolveBlockStatus(resolveStagingArea, resolveTip, true)
+	return nil, externalapi.StatusInvalid, nil
+}
+
+func (csm *consensusStateManager) getLowerTips(stagingArea *model.StagingArea, pendingTip *externalapi.DomainHash) ([]*externalapi.DomainHash, error) {
+	tips, err := csm.consensusStateStore.Tips(stagingArea, csm.databaseContext)
+	if err != nil {
+		return nil, err
+	}
+
+	lowerTips := []*externalapi.DomainHash{pendingTip}
+	for _, tip := range tips {
+		if tip.Equal(pendingTip) {
+			continue
+		}
+		selectedParent, err := csm.ghostdagManager.ChooseSelectedParent(stagingArea, tip, pendingTip)
 		if err != nil {
-			return nil, false, err
+			return nil, err
 		}
+		if selectedParent.Equal(pendingTip) {
+			lowerTips = append(lowerTips, tip)
+		}
+	}
+	return lowerTips, nil
+}
 
-		if blockStatus == externalapi.StatusUTXOValid {
-			selectedTip = resolveTip
-			isCompletelyResolved = !hasMoreUnverifiedThanMax
+func (csm *consensusStateManager) ResolveVirtual(maxBlocksToResolve uint64) (*externalapi.VirtualChangeSet, bool, error) {
+	onEnd := logger.LogAndMeasureExecutionTime(log, "csm.ResolveVirtual")
+	defer onEnd()
 
-			err = staging.CommitAllChanges(csm.databaseContext, resolveStagingArea)
-			if err != nil {
-				return nil, false, err
-			}
+	readStagingArea := model.NewStagingArea()
 
-			if reversalData != nil {
-				err = csm.ReverseUTXODiffs(resolveTip, reversalData)
-				if err != nil {
-					return nil, false, err
-				}
-			}
-			break
-		}
+	pendingTip, pendingTipStatus, err := csm.findNextPendingTip(readStagingArea)
+	if err != nil {
+		return nil, false, err
 	}
 
-	if selectedTip == nil {
+	if pendingTip == nil {
 		log.Warnf("Non of the DAG tips are valid")
 		return nil, true, nil
 	}
 
-	oldVirtualGHOSTDAGData, err := csm.ghostdagDataStore.Get(csm.databaseContext, readStagingArea, model.VirtualBlockHash, false)
+	prevVirtualSelectedParent, err := csm.virtualSelectedParent(readStagingArea)
 	if err != nil {
 		return nil, false, err
 	}
 
+	if pendingTipStatus == externalapi.StatusUTXOValid && prevVirtualSelectedParent.Equal(pendingTip) {
+		return nil, true, nil
+	}
+
+	// Resolve a chunk from the pending chain
+	resolveStagingArea := model.NewStagingArea()
+	unverifiedBlocks, err := csm.getUnverifiedChainBlocks(resolveStagingArea, pendingTip)
+	if err != nil {
+		return nil, false, err
+	}
+
+	intermediateTip := pendingTip
+
+	// Too many blocks to verify, so we only process a chunk and return
+	if maxBlocksToResolve != 0 && uint64(len(unverifiedBlocks)) > maxBlocksToResolve {
+
+		intermediateTipIndex := uint64(len(unverifiedBlocks)) - maxBlocksToResolve
+		intermediateTip = unverifiedBlocks[intermediateTipIndex]
```

Comment:
I don't like the name "intermediateTip" too much, first, because it's not really a tip, and second, because the word "intermediate" doesn't convey much info.

Consider renaming to "stopBlock", or "highestBlockToResolve" or something similar.